lib/tapioca/executor.rb

```diff
@@ -25,24 +26,54 @@ def initialize(queue, number_of_workers: nil)
       @items_per_worker = T.let((queue.length.to_f / @number_of_workers).ceil, Integer)
     end
 
-    sig { params(block: T.proc.params(item: T.untyped).void).void }
+    sig do
+      type_parameters(:T).params(
+        block: T.proc.params(item: T.untyped).returns(T.type_parameter(:T))
+      ).returns(T::Array[T.type_parameter(:T)])
+    end
     def run_in_parallel(&block)
       # If we only have one worker selected, it's not worth forking, just run sequentially
-      if @number_of_workers == 1
-        block.call(@queue.shift) until @queue.empty?
-        return
-      end
+      return @queue.map { |item| block.call(item) }.compact if @number_of_workers == 1
+
+      read_pipes = []
+      write_pipes = []
 
       # If we have more than one worker, fork the pool by shifting the expected number of items per worker from the
       # queue
       workers = (0...@number_of_workers).map do
         items = @queue.shift(@items_per_worker)
 
-        fork { block.call(items.shift) until items.empty? }
+        # Each worker has their own pair of pipes, so that we can read the result from each worker separately
+        read, write = IO.pipe
+        read_pipes << read
+        write_pipes << write
+
+        fork do
+          read.close
+          result = items.map { |item| block.call(item) }.compact
```

Comment:
Now that I think about this, maybe we should not be doing a `compact` here and just return an array of whatever the block call returned. There could be instances in the future where we might want to know what was returned even if it was `nil`. Just noting for the future, though, this is good as it stands now.